refactor(formatter): implement `GetSpan` for `AstNode` instead of generating a `span` method

Dunqing opened the pull request:
`GetSpan` can be used in a generic method that can constrain a generic parameter to satisfy `GetSpan` trait. So that would be useful than a pure method.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12423?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12424** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12424?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12423** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12423?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12423?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12422** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12422?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12421** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12421?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12420** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12420?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12356** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12356?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12423).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel reviewed (commented):
You could, if you wanted, make the generated code a bit shorter by eliding lifetimes and importing `oxc_span::Span` at top of file:

```rs
impl GetSpan for AstNode<'_, Program<'_>> {
    #[inline]
    fn span(&self) -> Span {
        self.inner.span()
    }
}
```

In `ast_tools` you can get type name with `'_`  lifetimes using `type_def.ty_anon(schema)`.

There's then only 1 other change you'd need to make to generated code to remove `#![expect(clippy::elidable_lifetime_names)]`:

```rs
impl<T: fmt::Debug> fmt::Debug for AstNode<'_, T> { /* ... */ }
```

It's not important, just letting you know that `ast_tools` has that method...

Dunqing commented:
> You could, if you wanted, make the generated code a bit shorter by eliding lifetimes and importing `oxc_span::Span` at top of file:
> 
> ```rust
> impl GetSpan for AstNode<'_, Program<'_>> {
>     #[inline]
>     fn span(&self) -> Span {
>         self.inner.span()
>     }
> }
> ```
> 
> In `ast_tools` you can get type name with `'_` lifetimes using `type_def.ty_anon(schema)`.
> 
> There's then only 1 other change you'd need to make to generated code to remove `#![expect(clippy::elidable_lifetime_names)]`:
> 
> ```rust
> impl<T: fmt::Debug> fmt::Debug for AstNode<'_, T> { /* ... */ }
> ```
> 
> It's not important, just letting you know that `ast_tools` has that method...

Thank you for looking into this! Yes, the generated code can be simplified a lot. The `GetSpan` implementation code that I copied from another place where in the same file, which can be simplified as well. I will simplify them in the follow-up PR according to your suggestion later.

Dunqing committed:
refactor(formatter): implement `GetSpan` for `AstNode` instead of generating a `span` method (#12423)

`GetSpan` can be used in a generic method that can constrain a generic parameter to satisfy `GetSpan` trait. So that would be useful than a pure method.

overlookmotel commented:
I only looked at it because I thought you'd requested my review. But of course actually it was an "auto review request" because the PR touches `ast_tools`!

If you do want to make changes to the codegen, the only suggestion I'd make is to move any static code out into a normal module, rather than codegen-ing it. e.g.:

https://github.com/oxc-project/oxc/blob/6838948e8d0907e90f05aede3d28c6fd7beba37b/tasks/ast_tools/src/generators/formatter/ast_nodes.rs#L179-L184

https://github.com/oxc-project/oxc/blob/6838948e8d0907e90f05aede3d28c6fd7beba37b/tasks/ast_tools/src/generators/formatter/ast_nodes.rs#L553-L600

i.e. only codegen code that *needs* to be codegen-ed because it relies on AST schema.

The (small) advantage is then it's a bit easier to edit this code because you get type annotations, auto-format etc in IDE, which you don't get inside `quote!`.

Disadvantage I guess is code ends up spread around a bit more. So pluses and minuses - feel free to ignore if you think it's better as it is.

Dunqing commented:
@overlookmotel Thanks! I got your meaning. While writing, I was considering whether this part should be moved to another module and implemented using `macro!`. Now I know this should be. I can see the only problem is that these implementations use internal fields marked as `pub(super)`, and I don't want to make it more public; anyway, it doesn't matter.
